File: 0234-palindrome-linked-list/0234-palindrome-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public boolean isPalindrome(ListNode head) {
        ListNode slow=head,fast=head,prev=null;
        while(fast!=null && fast.next!=null){
            slow=slow.next;
            fast=fast.next.next;
        }
        while(slow !=null){
            ListNode temp=slow.next;
            slow.next=prev;
            prev=slow;
            slow=temp;
        }
        ListNode current=head;
        while(prev!=null){
            if(prev.val!=current.val) return false;
            current=current.next;
            prev=prev.next;
        }
        return true;
    }
}

File: 0234-palindrome-linked-list/0234-palindrome-linked-list.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
bool isPalindrome(struct ListNode* head) {
    if (head == NULL || head->next == NULL) return true;
    struct ListNode *slow=head;
    struct ListNode *fast=head;
    while(fast!=NULL && fast->next!=NULL){
        slow=slow->next;
        fast=fast->next->next;
    }
    struct ListNode *prev=NULL;
    while(slow!=NULL){
        struct ListNode *temp=slow->next;
        slow->next=prev;
        prev=slow;
        slow=temp;
    }
    fast=head;
    while(prev!=NULL){
        if(prev->val!=fast->val) return false;
        prev=prev->next;
        fast=fast->next;
    }
    return true;
}